waspc/src/Wasp/Generator/Valid/Validator.hs

```diff
@@ -0,0 +1,76 @@
+module Wasp.Generator.Valid.Validator where
+
+import Data.Bifunctor (first)
+import Data.List (intercalate)
+import Data.List.NonEmpty (NonEmpty)
+import qualified Data.List.NonEmpty as NE
+import Data.Maybe (maybeToList)
+import qualified Validation as V
+import Wasp.Util (indent)
+
+type Validator input result = input -> Validation result
+
+type Validation result =
+  V.Validation (NonEmpty ValidationError) result
+
+data ValidationError = ValidationError
```

Comment:
I was tempted to consider doing this as `data ValidationError = ValidationError String | InField FieldName ValidationError | InFile FileName ValidationError`.

This then has some nice recursive properties. But, it also allows for too much freedom, so you could end up with InFile -> InField -> InFile -> InField -> message. This could be avoided by further refining the data types but then it gets complex, so I think it is best to stick with the solution you have. Just thinking out loud.